editors/code/src/test_explorer/discover_and_update.ts

```diff
@@ -0,0 +1,594 @@
+    updateTestItemsByModel();
+}
+
+function updateTestItemsByModel() {
+    testController!.items.replace([]);
+    const rootTestItems = VscodeTestTreeBuilder.build();
+    // const mockItem = testController!.createTestItem('123', '456');
+    // testController!.items.add(mockItem);
+    testController!.items.replace(rootTestItems);
+}
+
+async function updateModelByChangeOfFile(uri: vscode.Uri) {
+    const testInfos = await RaApiHelper.getTestRunnablesInFile(uri);
+
+    // Maybe from some to none
+    // need to recursively clean the parent, until there is at least one test cases.
+    if (testInfos === null || testInfos.length === 0) {
+        testModelTree.removeTestItemsRecusivelyByUri(uri);
+        return;
+    }
+
+    // Ensure parent is created in the test-model-tree
+    const runnables = testInfos.map(it => new RunnableFacde(it.runnable));
+
+    const testModuelRunnables = runnables.filter(it =>
+        it.testKind === NodeKind.TestModule)
+        .sort(RunnableFacde.sortByLabel);
+
+    const testItemRunnables = runnables.filter(it =>
+        it.testKind === NodeKind.Test);
+
+    assert(testModuelRunnables.length + testItemRunnables.length === runnables.length);
+
+    const rootTestModuleRunnbale = testModuelRunnables[0];
+
+    // Now, we know the root test module
+    // Note that the parent might be not exist in the tree.
+    // Then how to update the test model tree?
+    // There are two ways:
+    // - down-to-up, which means get parent module,  and create parent recursively
+    // - up-to-down, find the nearest parent, and create children for it
+    // up-to-down is simpler, since the logic could be reused,
+    // for anyway we need to init, which is up-to-dowm :P
+    // PS: we could set dirty flag to make it more effective, but I think it's fine
+
+    // Observe that:
+    // if node is new in the tree, fetch its children too
+    // if node is existing in the tree,
+    // - and it's definition, then we have got all of the runnables(because they
+    // are in same file) and will be updated
+    // - and it's declaration
+    //   - and definition uri is not changed, then not fetch children
+    //   - and definition uri is changed, then fetch the children
+
+    // So, the final steps are:
+    // 1. Make the change(remove, add new test module declaration or add test module with items),
+    // but not fetch children for new test module declaration
+    // 2. collectt which children need to be fetched(reuse logic)
+    // 3. fetch children(reuse logic)
+
+    const nearestNode = testModelTree.findNearestNodeByRunnable(rootTestModuleRunnbale);
+
+    assert(nearestNode.kind !== NodeKind.Test, "the nearest node of root module can not be a test node.");
+
+    // create target node
+    // This is necessary, because we do not know how many targets a package contains unless we fetch data of `cargo metadata`
+    // But we want to only fetch it when cargo file is changed, to make things more lazily.
```

Comment:
I'm not sure if this is strictly necessary, as a change in the Cargo.toml will be picked up rust-analyzer-the-server, and querying the language server to get the new workspace structure feels like a far more scalable and reasonable solution than attempting to figure this out in the client.

Reply:
It's only necessary for current implmentation.

We maintain a test model tree, and build test-item tree from test model tree.

For now, we only create test item for target when it does have test children. Which means, for target that not have test children, we remove it from our test model tree. So, when a new test is added to a file, the test might not have a valid target in the test model tree. We add the target node back to test model tree here.

Of course, we could choose to refactor the logic a bit, like, keep target nodes forever and change the logic how we rebuild test-item tree from test model tree.